### DISCUSSION TITLE
关于第三课<通过函数提取合约中的ETH>  refund 方法的疑问

### DISCUSSION THREAD

**Discussion by mi1chstrasse** [smartcontractkit/Web3_tutorial_Chinese#1] [General]
```solidity
    function refund() external windowClosed {
        require(convertEthToUsd(address(this).balance) < TARGET, "Target is reached");
        require(fundersToAmount[msg.sender] != 0, "there is no fund for you");
        bool success;
        (success, ) = payable(msg.sender).call{value: fundersToAmount[msg.sender]}("");
        require(success, "transfer tx failed");
        fundersToAmount[msg.sender] = 0;
    }
```
此处回退为何不消耗 gas
按视频中的测试，两个账户均发送了0.1 个 eth，回退均得到了 0.1 个 eth。
此时为何不消耗 gas。
按我的理解此地址总转出金额应该小于 0.2eth

**Comment by QingyangKong** [comment]
因为gas fee是由交易的sender再支付，所以不会在 transfer 的金额中被扣除

**Reply by heisedeheiheihei** [reply]
是否可以这样理解：发送的eth和gas fee是独立的两个部分，发送/接收的数额是确定的，gas fee是为了完成交易而计算出来的独立于商品价格的”手续费“，由sender支付，且仅在交易完成时扣除，如果发送了回退，整个交易未完成，原数额返回给sender，费用也不再需要支付

**Reply by mi1chstrasse** [reply]
明白了

> 是否可以这样理解：发送的eth和gas fee是独立的两个部分，发送/接收的数额是确定的，gas fee是为了完成交易而计算出来的独立于商品价格的”手续费“，由sender支付，且仅在交易完成时扣除，如果发送了回退，整个交易未完成，原数额返回给sender，费用也不再需要支付



> 是否可以这样理解：发送的eth和gas fee是独立的两个部分，发送/接收的数额是确定的，gas fee是为了完成交易而计算出来的独立于商品价格的”手续费“，由sender支付，且仅在交易完成时扣除，如果发送了回退，整个交易未完成，原数额返回给sender，费用也不再需要支付

交易发生了，必然会支付gas fee，与交易成功或失败没有关系，因为即使交易失败，矿工依旧为此交易进行校验和计算，消耗了资源。

**Comment by MrCare** [comment]
请问这一块在最后清零会不会有被重入攻击的风险？
